sdk/go/pulumi/context.go

```diff
@@ -806,6 +806,20 @@ func (ctx *Context) registerResource(
 	parent := options.Parent
 	if options.Parent == nil {
 		options.Parent = ctx.stack
+	} else {
+		// Guard against uninitialized parent resources to prevent
+		// panics from invalid state further down the line.
+		// Uninitialized parent resources won't have a URN.
+		if parent.URN().getState() == nil {
+			var msg string
+			if _, parentIsCustom := parent.(CustomResource); !parentIsCustom {
+				msg = "parent component resource %T has not been registered: " +
+					"did you mean to call RegisterComponentResource?"
+			} else {
+				msg = "parent resource %T has not been registered"
```

Comment:
This should probably match the phrasing above, and say that RegisterResource should have been called on the parent.

Reply:
I considered that, and I'm down to add that.
The reason I didn't at the time was that custom resources typically have a NewMyResource constructor in our generated SDKs, so end users won't _normally_ call RegisterResource directly.
If it's fine to leak that little detail, I can fix that.

Reply:
hmm maybe that they didn't call the "NewX" constructor method then?

Reply:
AFAIK, the SDK doesn't know that NewX exists. I'm going to stick with RegisterResource for now, since this variant of the error is less likely to happen (passing in a zero value `s3.Bucket` instead of one returned by `s3.NewBucket`), so the little leak that all resource instantiations go through RegisterResource isn't too bad.
(If there's a type-constructor mapping, though, we can definitely tell users that they forgot to call NewFoo.)